contracts/NodeOperatorRegistry.sol

```diff
@@ -0,0 +1,642 @@
+            "Invalid amount"
+        );
+        require(
+            _heimdallFee >= state.minHeimdallFees &&
+                _heimdallFee <= state.maxHeimdallFees,
+            "Invalid heimdallFee"
+        );
+
+        uint256 operatorId = operatorOwners[msg.sender];
+        require(operatorId != 0, "Operator not exists");
+
+        Operator.NodeOperator storage no = operators[operatorId];
+        require(
+            no.status == Operator.NodeOperatorStatus.ACTIVE,
+            "The Operator status is not active"
+        );
+
+        // stake a validator
+        (no.validatorId, no.validatorShare) = IValidator(no.validatorContract)
+            .stake(
+                msg.sender,
+                _amount,
+                _heimdallFee,
+                true,
+                no.signerPubkey,
+                no.commissionRate
+            );
+
+        // update the operator status to STAKED.
+        no.status = Operator.NodeOperatorStatus.STAKED;
+        no.commissionRate = state.commissionRate;
+
+        // update global state.
+        state.totalActiveNodeOpearator--;
+        state.totalStakedNodeOpearator++;
+
+        emit StakeOperator(operatorId, no.validatorId);
+    }
+
+    /// @notice Alloxs to restake Matics to Polygin stakeManager
+    /// @param _amount amount to stake.
+    function restake(uint256 _amount) external override whenNotPaused {
+        require(state.restake, "Restake is disabled");
+
+        uint256 operatorId = operatorOwners[msg.sender];
+        require(operatorId != 0, "Operator not exists");
+
+        Operator.NodeOperator storage no = operators[operatorId];
+        require(
+            no.status == Operator.NodeOperatorStatus.STAKED,
+            "The operator status is not STAKED"
+        );
+
+        require(_amount > 0, "Amount is ZERO");
+
+        IValidator(no.validatorContract).restake(
+            msg.sender,
+            no.validatorId,
+            _amount,
+            false
+        );
+
+        emit RestakeOperator(operatorId, no.validatorId);
+    }
+
+    /// @notice Unstake a validator from the Polygon stakeManager contract.
+    function unstake() external override whenNotPaused {
+        uint256 id = operatorOwners[msg.sender];
+        require(id != 0, "Operator not exists");
+
+        Operator.NodeOperator storage no = operators[id];
+        require(
+            no.status == Operator.NodeOperatorStatus.STAKED,
+            "The operator status is not STAKED"
+        );
+        IValidator(no.validatorContract).unstake(no.validatorId);
+
+        no.status = Operator.NodeOperatorStatus.UNSTAKED;
+        state.totalStakedNodeOpearator--;
+        state.totalUnstakedNodeOpearator++;
+
+        emit UnstakeOperator(id);
+    }
+
+    /// @notice Allows to top up heimdall fees.
+    /// @param _heimdallFee amount
+    function topUpForFee(uint256 _heimdallFee) external override whenNotPaused {
+        require(_heimdallFee > 0, "HeimdallFee is ZERO");
```

Comment:
Should it also be checked that id doesnt go over heimdall fee max?

Reply:
Yes, I'll update